File: fix_server.py
with open("server.ts", "r") as f:
    content = f.read()

newsletter_block = """
  app.post("/api/newsletter/subscribe", async (req, res) => {
    try {
      const { email } = req.body;
      if (!email) return res.status(400).json({ error: "Email is required" });
      const existing = await prisma.newsletterSubscriber.findUnique({ where: { email } });
      if (existing) {
        return res.json({ message: "Already subscribed", subscriber: existing });
      }
      const subscriber = await prisma.newsletterSubscriber.create({ data: { email } });
      res.json({ message: "Subscribed successfully", subscriber });
    } catch (error) {
      console.error(error);
      res.status(500).json({ error: "Failed to subscribe" });
    }
  });
"""

# Remove the incorrectly placed one
content = content.replace(newsletter_block, "")
# And also the extra blank lines if any, but replacing the block itself is enough.

# Now place it BEFORE Vite middleware / catch-all
insertion = """
  // Newsletter
""" + newsletter_block

content = content.replace("  // Vite middleware for development", insertion + "\n  // Vite middleware for development")

with open("server.ts", "w") as f:
    f.write(content)
print("Fixed server.ts")

File: fix_server.js
import fs from 'fs';
let code = fs.readFileSync('server.ts', 'utf8');

const mwCode = `
  const editorOrAdminMiddleware = async (req: any, res: any, next: any) => {
    authMiddleware(req, res, () => {
      if (req.user.role !== "ADMIN" && req.user.role !== "EDITOR")
        return res.status(403).json({ error: "Forbidden" });
      next();
    });
  };

  const adminMiddleware = async (req: any, res: any, next: any) => {
    authMiddleware(req, res, () => {
      if (req.user.role !== "ADMIN")
        return res.status(403).json({ error: "Forbidden" });
      next();
    });
  };
`;

code = code.replace(/  const editorOrAdminMiddleware = async \(req: any, res: any, next: any\) => \{[\s\S]*?  const adminMiddleware = async \(req: any, res: any, next: any\) => \{[\s\S]*?    \}\);\n  \};\n/m, '');

code = code.replace(
  /  const authMiddleware = async \(req: any, res: any, next: any\) => \{[\s\S]*?    \}\n  \};\n/,
  match => match + mwCode
);

fs.writeFileSync('server.ts', code);
